app/src/main/java/com/example/booklisting/MainActivity.java

```diff
@@ -130,38 +125,42 @@ public void onLoadMore() {
     }
 
     @Override
-    public void onLoadFinished(@NonNull Loader<List<Book>> loader, List<Book> list) {
+    public void onLoadFinished(@NonNull Loader<BookLoaderResult> loader, BookLoaderResult loaderResult) {
         if ((adapter == null) || (adapter.getItemCount() == 0)) {
-            if (exceptionThrown) {
+            if (loaderResult.getException() != null) {
+                Log.i("Exception is", String.valueOf(loaderResult.getException()));
                 progressBar.setVisibility(View.GONE);
                 emptyView.setVisibility(View.GONE);
                 showAlert();
             } else {
 
-                if (list.isEmpty()) {
+                if (loaderResult.getResult().isEmpty()) {
```

Comment:
Are you sure there is not possibility of NPE here? (NullPointerException). If there is a error, result array inside loaderResult could be null if I'm not mistaken?

Reply:
You`re right, I missed that. I added another check at the beginning of onLoadFinished() method.